runners/spark/src/main/java/org/apache/beam/runners/spark/SparkPipelineResult.java

```diff
@@ -207,12 +202,19 @@ protected State awaitTermination(final Duration duration) throws ExecutionExcept
            terminationState = State.DONE;
            break;
          default:
-           state = State.UNKNOWN;
+           terminationState = State.UNKNOWN;
            break;
        }
        return terminationState;
     }
+  }
 
+  private void changeState(State newState) {
```

Comment:
I made all mutations of state in the class use this method, so it is for changing the state.
It also calls `stop()` when a state is changed from a non terminal state to a terminal state, as you said.